### PR TITLE
fix test teardown race with ongoing test operation

### PR DESCRIPTION
**Pull request by temichus**
This commit brings several new features in scylla_cluster.py to fix runaway asyncio task problems in topology tests  

-  Start-Stop Lock and Stop Event in ScyllaServer
- Tasks History, Wait for tasks from Tasks History and Manager broken state in ScyllaClusterManager
- make ManagerClient object function scope
- test_finished_event in ManagerClient

Fixes: scylladb/scylladb#16472
Fixes: scylladb/scylladb#16651

### CODE REVIEW

**Comment by temichus** [comment]
tested with following reproducers:
https://github.com/scylladb/scylladb/issues/16651#issue-2066160420

and  https://github.com/scylladb/scylladb/issues/16472#issuecomment-1910125573

File: test/pylib/scylla_cluster.py [hunk 1]
```diff
@@ -200,6 +201,15 @@ class CqlUpState(Enum):
     CONNECTED = 2,
     QUERIED = 3
 
+
+def start_stop_lock(func):
+    async def wrap(self, *args, **kwargs):
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:206]: please add type hint to `self`: `self: ScyllaServer`

File: test/pylib/scylla_cluster.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 """Scylla clusters for testing.
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:6]: This should be three separate commits:
- test/pylib: scylla_cluster: introduce ScyllaServer start-stop lock
- test/pylib: scylla_cluster: introduce task history
- test/pylib: scylla_cluster: fix teardown race with ongoing test operation

(and in last commit, explain how the lock and the history are used to fix the race)

File: test/pylib/scylla_cluster.py [hunk 3]
```diff
@@ -1164,11 +1185,19 @@ async def catching_handler(request) -> aiohttp.web.Response:
                     return aiohttp.web.Response(status=500, text=str(e))
             return catching_handler
 
+        def route_history_wrapper(handler: Callable)-> Callable:
+            @wraps(handler)
+            async def inner_wrapper(request):
+                self.logger.info("[ScyllaClusterManager][%s] %s", asyncio.current_task().get_name(), request.url)
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:1191]: Can you provide sample output from test.py logs, how it looks like with this change?

File: test/pylib/scylla_cluster.py [hunk 4]
```diff
@@ -1249,6 +1278,21 @@ async def _after_test(self, _request) -> str:
             self.current_test_case_full_name = ''
         self.is_after_test_ok = True
         cluster_str = str(self.cluster)
+
+        # cancel all other tasks in ScyllaClusterManager
+        self.logger.info(self.repr_tasks_history())
+        for key, value in self.tasks_history.items():
+            if "after-test" not in value.path_qs:
+                key.cancel(msg="test finished")
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:1286]: Shouldn't we wait for tasks to finish instead of cancelling them? (Not sure what cancel does though)

File: test/pylib/scylla_cluster.py [hunk 5]
```diff
@@ -1249,6 +1278,21 @@ async def _after_test(self, _request) -> str:
             self.current_test_case_full_name = ''
         self.is_after_test_ok = True
         cluster_str = str(self.cluster)
+
+        # cancel all other tasks in ScyllaClusterManager
+        self.logger.info(self.repr_tasks_history())
+        for key, value in self.tasks_history.items():
+            if "after-test" not in value.path_qs:
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:1285]: I think the implementation of `_after_test` should not depend on how the route that calls `_after_test` is named -- it's an abstraction leak, dependency reversal.

File: test/pylib/scylla_cluster.py [hunk 6]
```diff
@@ -1249,6 +1278,21 @@ async def _after_test(self, _request) -> str:
...
+                key.cancel(msg="test finished")
+        for key, value in self.tasks_history.items():
+            if not key.done() and "after-test" not in value.path_qs:
+                self.logger.info("wait for cancelling task:%s, request:%s", key.get_name(), value.path_qs)
+                try:
+                    await key
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:1291]: I wonder if we should introduce a timeout here and kill everything if it doesn't finish after timeout.

(But we do need to make sure to cleanup resources -- I don't know if some of these tasks might be responsible for that)

**Review by kbr-scylla — CHANGES_REQUESTED** [review]

File: test/pylib/scylla_cluster.py (hunk 3 above)
Comment (temichus) [reply test/pylib/scylla_cluster.py:1191]: there is some duplication on the logs (with aiohttp lib). but now I can see the task name and request a connection for better troubleshooting
```
12:16:25.412 INFO>  [18/Apr/2024:11:16:18 +0100] "PUT /cluster/addservers HTTP/1.1" 200 529 "-" "Python/3.11 aiohttp/3.9.3"
12:16:25.419 INFO> [release/topology.test_repr.1] [ScyllaClusterManager][Task-166] http://api/cluster/server/4/get_log_filename
12:16:25.419 INFO>  [18/Apr/2024:11:16:25 +0100] "GET /cluster/server/4/get_log_filename HTTP/1.1" 200 230 "-" "Python/3.11 aiohttp/3.9.3"
12:16:25.424 INFO> [release/topology.test_repr.1] [ScyllaClusterManager][Task-175] http://api/cluster/server/5/get_log_filename
12:16:25.425 INFO>  [18/Apr/2024:11:16:25 +0100] "GET /cluster/server/5/get_log_filename HTTP/1.1" 200 230 "-" "Python/3.11 aiohttp/3.9.3"
12:16:25.425 INFO> [release/topology.test_repr.1] [ScyllaClusterManager][Task-176] http://api/cluster/server/6/get_log_filename
12:16:25.426 INFO>  [18/Apr/2024:11:16:25 +0100] "GET /cluster/server/6/get_log_filename HTTP/1.1" 200 230 "-" "Python/3.11 aiohttp/3.9.3"
12:16:25.426 INFO> [release/topology.test_repr.1] [ScyllaClusterManager][Task-177] http://api/cluster/server/7/get_log_filename
12:16:25.426 INFO>  [18/Apr/2024:11:16:25 +0100] "GET /cluster/server/7/get_log_filename HTTP/1.1" 200 230 "-" "Python/3.11 aiohttp/3.9.3"
12:16:25.427 INFO> [release/topology.test_repr.1] [ScyllaClusterManager][Task-178] http://api/cluster/server/8/get_log_filename
12:16:25.427 INFO>  [18/Apr/2024:11:16:25 +0100] "GET /cluster/server/8/get_log_filename HTTP/1.1" 200 230 "-" "Python/3.11 aiohttp/3.9.3"
12:16:25.429 INFO> [release/topology.test_repr.1] [ScyllaClusterManager][Task-179] http://api/cluster/server/4/stop_gracefully
```

File: test/pylib/scylla_cluster.py (hunk 6 above)
Comment (temichus) [reply test/pylib/scylla_cluster.py:1291]: Asyncio tasks cannot be killed, only cancelled, even cancel is now guaranteed that the task is finished immediately.

already wrote about it here https://github.com/scylladb/scylladb/issues/16472#issuecomment-1898807664

Comment (temichus) [reply test/pylib/scylla_cluster.py:1291]: regarding cleanup and timeout - will do additional testing

File: test/pylib/scylla_cluster.py (hunk 4 above)
Comment (temichus) [reply test/pylib/scylla_cluster.py:1286]: it can be too long. For example in reproducer  https://github.com/scylladb/scylladb/issues/16651#issue-2066160420

I faced an issue that waiting node_start took TOPOLOGY_TIMEOUT = 1000 sec. I think it is a quite big timeout especially if we are in _after_test. meaning the test has already been finished

**Comment by kostja** [comment]
@kbr-scylla I am not a maintainer here but shouldn't the two independent problems addreesed in the patch be split into two separate commits?

File: test/pylib/scylla_cluster.py (hunk 2 above)
Comment (temichus) [reply test/pylib/scylla_cluster.py:6]: Could you please describe why?

because this 2 logic can not work separately 

with start-stop lock without tasks cancelling/awaiting - we will still faced with issues teardown
and vice versa is the same

Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:6]: You don't need to solve all problems in a single commit.

You can use one commit to introduce some concept which will be useful later to solve a problem, then another commit to introduce yet another concept, then another commit that uses the previously introduced concepts to actually solve some problem.

Generally we try to split commits in scylladb repo on module or feature boundaries. If you introduce two modules or two features, they should be in separate commits.

This helps reviewers keep track of what's happening.

Otherwise when I review one large commit and read it top to bottom I see:
- fragment of code related to feature/concept X
- then fragment of code related to feature/concept Y
- then fragment related to X again

etc. and it's hard for me to understand everything because I have limited memory in my head.

In your commit message you even have two headers:
"Introducing the Start-Stop Lock:"
and
"Introducing Tasks History:"

so these two concepts can definitely be understood independently of each other.

Then you connect two separate concepts so they work together to solve a problem.

But they are still separate concepts. So they should be introduced in separate commits.

Comment (temichus) [reply test/pylib/scylla_cluster.py:6]: split in 2. not sure what code change needs to be added in 3rd commit

File: test/pylib/scylla_cluster.py [hunk 7]
```diff
@@ -1164,11 +1185,20 @@ async def catching_handler(request) -> aiohttp.web.Response:
                     return aiohttp.web.Response(status=500, text=str(e))
             return catching_handler
 
+        def route_history_wrapper(handler: Callable)-> Callable:
+            @wraps(handler)
+            async def inner_wrapper(request):
+                self.logger.info("[ScyllaClusterManager][%s] %s", asyncio.current_task().get_name(), request.url)
+                self.tasks_history[asyncio.current_task()] = request
+                ret = await handler(request)
+                return ret
```
[the remark this answers is not in the document]
Comment (temichus) [reply test/pylib/scylla_cluster.py]: fixed

File: test/pylib/scylla_cluster.py (hunk 1 above)
Comment (temichus) [reply test/pylib/scylla_cluster.py]: fixed

File: test/pylib/scylla_cluster.py (hunk 5 above)
Comment (temichus) [reply test/pylib/scylla_cluster.py]: logic changed, now I'm looking into task name

File: test/pylib/scylla_cluster.py (hunk 6 above)
Comment (temichus) [reply test/pylib/scylla_cluster.py]: added timeout and waited for PUT actions in no raise an error

regarding cleanup. 
as I mentioned is https://github.com/scylladb/scylladb/issues/16472#issuecomment-1898807664 
Asyncio tasks cannot be killed, only finished

File: test/pylib/scylla_cluster.py [hunk 8]
```diff
@@ -1240,6 +1294,28 @@ async def _before_test_req(self, request) -> str:
...
+                    await asyncio.wait_for(task, timeout=120)
+        except asyncio.TimeoutError:
+            self.logger.error(f"error on waiting coro {task.get_name()}")
+            sys.exit(1)
+
+        # check on tasks leakage
+        await asyncio.sleep(0.1)
+        if self.tasks_history:
+            self.logger.error("tasks leakage found %s, STOP PYTEST" ,self.tasks_history)
+            sys.exit(1)
```
Comment (temichus) [inline test/pylib/scylla_cluster.py:1317]: @kbr-scylla @kostja 

Bad news,
Waiting for all tasks does not guarantee that lesf tasks in asyncio.gather will not spawn new tasks while we waiting and 
 sys.exit(1) in pytest crash test.py. itself 
So, more logic is coming

I really suggest to use https://github.com/scylladb/scylladb/pull/17781  instead all of this new complicated logic

PR https://github.com/scylladb/scylladb/pull/17781  contains only 8 lines of new logic (other in renaming and doc strings)
rather than this PR has already 100 lines of new logic without 100% guarantee that ig handle all cases of wrong asyncio.gather usages in tests/test.py

cc @xtrey

Comment (temichus) [reply test/pylib/scylla_cluster.py]: looks like we safe now

File: test/pylib/scylla_cluster.py [hunk 9]
```diff
@@ -1301,12 +1306,19 @@ async def _after_test(self, _request) -> str:
...
+                if not task.done():
+                    self.logger.info("wait for task:%s, request:%s", task, request.path_qs)
+                    await asyncio.wait_for(task, timeout=120)
+        except asyncio.TimeoutError:
+            self.break_manager(f"error on waiting coro {task.get_name()}")
+
+        # check on tasks leakage
+        await asyncio.sleep(0.1)
+        if self.tasks_history:
+            self.break_manager(f"tasks leakage found  {self.tasks_history}")
```
Comment (piodul) [inline test/pylib/scylla_cluster.py:1321]: This solution doesn't look reliable... You can always have a runaway asyncio task which waits for a long time and then sends some requests. I don't think this will catch it.

I think it would be better if every test case received its own instance of the ManagerClient (as far as I know right now we are reusing it for multiple test cases). At the end of the test, we could gracefully stop it, preventing more requests from being sent and waiting for the existing ones to complete.

If sharing some resources (e.g. the HTTP client) is important, then the instances of the manager client should be able to do it.

I know that, technically speaking, a test case could create its own manager instance by itself and this mechanism wouldn't catch this. However, I believe that such mistake is much less likely to happen and is much more likely to be spotted.

@kostja , @kbr-scylla - do you perhaps know if there is a good reason why we reuse the same ManagerClient instance for multiple tests?

Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:1321]: > @kostja , @kbr-scylla - do you perhaps know if there is a good reason why we reuse the same ManagerClient instance for multiple tests?

ManagerClient corresponds to a ScyllaCluster (1-1 correspondence) and clusters can in theory be reused between test cases (if they are not left dirty and are in the same suite)

File: test/pylib/scylla_cluster.py [hunk 10]
```diff
@@ -566,6 +583,7 @@ async def shutdown_control_connection(self) -> None:
             self.control_cluster.shutdown()
             self.control_cluster = None
 
+    @stop_event
     @start_stop_lock
     async def stop(self) -> None:
```
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py]: IIUC the decorator applies to the thing below it
so `stop_event` is applied to (`start_stop_lock` applied to `stop`)

**Comment by kbr-scylla** [comment]
Also nit: commit messages should be prefixed with the name of the module they're modifying, for example
`test/pylib: Introduce Start-Stop Lock`

(we have no concrete rule what exactly the prefix should be, but look at the repo git log for inspiration or just use `test/pylib:`)

File: test/pylib/scylla_cluster.py (hunk 9 above)
Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:1321]: > ManagerClient corresponds to a ScyllaCluster (1-1 correspondence) and clusters can in theory be reused between test cases (if they are not left dirty and are in the same suite)

Hm, but that doesn't explain it. I'm a bit confused as well.

The scope of `manager` fixture is `function`:
```python3
@pytest.fixture(scope="function")
async def manager(request, manager_internal):
```

internally it uses `manager_internal` with the scope `session`
```python3
@pytest.fixture(scope="session")
async def manager_internal(event_loop, request):
```

IIUC we have different `pytest` executions for different test files, i.e., different test files == different sessions.

But clusters can be reused between different test files (?), since they are all accessing the same pool living in test.py process.

So different `ManagerClient`s can be used to reuse a cluster. (Need to confirm with experiment that it is indeed the case.)

So we should be able to change scope of `manager_internal` to be `function` and then we would have separate `ManagerClient`s.

But that is only the client-side, i.e. the `pytest` process. While the problematic part is the server-side, `ScyllaClusterManager` living inside the `test.py` process. The code we're looking at here is the server-side. Could we move all of this logic to the client-side instead (`ManagerClient`)?

If not, we'd also have to change `ScyllaClusterManager`, which IIUC currently is also per test file, not per test case.

That would multiply the amount of resources (like sockets), but also perhaps allow us to run test cases in the same test file concurrently. (Which is a long-requested feature.)

Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:1321]: > If not, we'd also have to change ScyllaClusterManager, which IIUC currently is also per test file, not per test case.

This'd have to be done anyway if we made `ManagerClient` function-scope, because two `ManagerClient`s cannot use a single `ScyllaClusterManager` concurrently.

File: test/pylib/scylla_cluster.py [hunk 11]
```diff
@@ -516,8 +548,10 @@ def report_error(message: str):
 
             # Sleep and retry
             await asyncio.sleep(sleep_interval)
-
-        report_error('failed to start the node, timeout reached')
+        reason = "timeout"
+        if self.stop_event.is_set():
+            reason= "stop_event"
```
[the remark this answers is not in the document]
Comment (temichus) [reply test/pylib/scylla_cluster.py]: outdated. 

```
reason = "timeout"
        if self.stop_event.is_set():
            reason = "stop_event"
```

File: test/pylib/scylla_cluster.py [hunk 12]
```diff
@@ -1204,11 +1212,19 @@ async def catching_handler(request) -> aiohttp.web.Response:
                     return aiohttp.web.Response(status=500, text=str(e))
             return catching_handler
 
+        def route_history_wrapper(handler: Callable)-> Callable:
+                @wraps(handler)
+                async def inner_wrapper(request):
+                    self.logger.info("[ScyllaClusterManager][%s] %s", asyncio.current_task().get_name(), request.url)
+                    self.tasks_history[asyncio.current_task()] = request
+                    return await handler(request)
+                return inner_wrapper
```
[the remark this answers is not in the document]
Comment (temichus) [reply test/pylib/scylla_cluster.py]: no, 4 space as needed

[the remark this answers is not in the document]
Comment (piodul) [reply test/pylib/scylla_cluster.py]: Ok, I did not select one line before that so maybe it's invisible:

```python
        def route_history_wrapper(handler: Callable)-> Callable:
                @wraps(handler)
                async def inner_wrapper(request):
                    self.logger.info("[ScyllaClusterManager][%s] %s", asyncio.current_task().get_name(), request.url)
                    self.tasks_history[asyncio.current_task()] = request
                    return await handler(request)
                return inner_wrapper
```

The `@wraps(handler)` line seems to be indented twice. There is a difference of 8 spaces with the previous line.

File: test/pylib/scylla_cluster.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (temichus) [reply test/pylib/scylla_cluster.py]: can't see where you found it 
```
reason = "timeout"
        if self.stop_event.is_set():
            reason = "stop_event"
```

[the remark this answers is not in the document]
Comment (piodul) [reply test/pylib/scylla_cluster.py]: Yes, I left this comment a long time ago - I only clicked "Review changes" today so you must have fixed it in the meantime.

File: test/pylib/scylla_cluster.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (temichus) [reply test/pylib/scylla_cluster.py]: yes start_stop_lock applied first

File: test/pylib/scylla_cluster.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (temichus) [reply test/pylib/scylla_cluster.py]: fixed

File: test/pylib/scylla_cluster.py [hunk 13]
```diff
@@ -535,8 +550,10 @@ def report_error(message: str):
 
             # Sleep and retry
             await asyncio.sleep(sleep_interval)
-
-        report_error('failed to start the node, timeout reached')
+        reason = "timeout"
+        if self.stop_event.is_set():
+            reason = "stop_event"
+        report_error(f'failed to start the node, {reason} reached')
```
[the remark this answers is not in the document]
Comment (temichus) [reply test/pylib/scylla_cluster.py]: fixed

**Comment by temichus** [comment]
> Also nit: commit messages should be prefixed with the name of the module they're modifying, for example `test/pylib: Introduce Start-Stop Lock`
> 
> (we have no concrete rule what exactly the prefix should be, but look at the repo git log for inspiration or just use `test/pylib:`)

fixed

File: test/pylib/scylla_cluster.py (hunk 9 above)
Comment (temichus) [reply test/pylib/scylla_cluster.py:1322]: @kbr-scylla @piodul 
client side fix for case of 
``` having a runaway asyncio task which waits for a long time and then sends some requests.```
https://github.com/scylladb/scylladb/pull/18563

Comment (temichus) [reply test/pylib/scylla_cluster.py:1315]: added 2 more commits to fixe described problem

**Review by kbr-scylla — APPROVED** [review]
PR description needs to be updated. Other than that LGTM.

I'm a bit worried with using multiple `ManagerClient`s that connect to a single `ScyllaClusterManager`, could this be a problem? Probably not, because currently test cases in a single file are running sequentially, we should not have a situation where two `ManagerClient`s are using the same `Manager` at the same time?

**Comment by temichus** [comment]
> PR description needs to be updated. Other than that LGTM.
> 

updated

> I'm a bit worried with using multiple `ManagerClient`s that connect to a single `ScyllaClusterManager`, could this be a problem? Probably not, because currently test cases in a single file are running sequentially, we should not have a situation where two `ManagerClient`s are using the same `Manager` at the same time?

that is correct. we create one Scylla Manager object per TopologyTest.run in test.py(that means per pytest run)
in one pytest run, we run tests sequentially with creating a new  ManagerClient connection before the test and closing it after the test.

**Comment by kbr-scylla** [comment]
Changed Fixes markers in commit message to:
```
Fixes: scylladb/scylladb#16472
Fixes: scylladb/scylladb#16651
```

(so bot can correctly react)

**Review by xtrey — APPROVED** [review]
LGTM
